packages/go_router/example/lib/stateful_nested_navigation.dart

```diff
@@ -0,0 +1,402 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/cupertino.dart';
+import 'package:flutter/material.dart';
+import 'package:go_router/go_router.dart';
+
+final GlobalKey<NavigatorState> _sectionANavigatorKey =
+    GlobalKey<NavigatorState>(debugLabel: 'sectionANav');
+final GlobalKey<NavigatorState> _sectionBNavigatorKey =
+    GlobalKey<NavigatorState>(debugLabel: 'sectionBNav');
+
+// This example demonstrates how to setup nested navigation using a
+// BottomNavigationBar, where each tab uses its own persistent navigator, i.e.
+// navigation state is maintained separately for each tab. This setup also
+// enables deep linking into nested pages.
+//
+// This example demonstrates how to display routes within a ShellRoute using a
+// `nestedNavigationBuilder`. Navigators for the tabs ('Section A' and
+// 'Section B') are created via nested ShellRoutes. Note that no navigator will
+// be created by the "top" ShellRoute. This example is similar to the ShellRoute
+// example, but differs in that it is able to maintain the navigation state of
+// each tab.
+
+void main() {
+  runApp(NestedTabNavigationExampleApp());
+}
+
+/// Extension to get the child widget of the Page. Wish there were a better way...
+extension PageWithChild on Page<dynamic> {
+  /// Gets the child of the page
+  Widget? findTheChild() {
+    final Page<dynamic> widget = this;
+    if (widget is MaterialPage) {
+      return widget.child;
+    } else if (widget is CupertinoPage) {
+      return widget.child;
+    } else if (widget is CustomTransitionPage) {
+      return widget.child;
+    }
+    return null;
+    // An unsafer option... :
+    //return (this as dynamic).child as Widget;
+  }
+}
+
+/// ShellRoute that uses a bottom tab navigation (ScaffoldWithNavBar) with
+/// separate navigators for each tab.
+class BottomTabBarShellRoute extends ShellRoute {
+  /// Constructs a BottomTabBarShellRoute
+  BottomTabBarShellRoute({
+    required this.tabs,
+    GlobalKey<NavigatorState>? navigatorKey,
+    List<RouteBase> routes = const <RouteBase>[],
+    Key? scaffoldKey = const ValueKey<String>('ScaffoldWithNavBar'),
+  }) : super(
+            navigatorKey: navigatorKey,
+            routes: routes,
+            nestedNavigationBuilder: (BuildContext context, GoRouterState state,
+                List<Page<dynamic>> pagesForCurrentRoute) {
+              // The first (and only) page will be the nested navigator for the
+              // current tab. The pages parameter will in this case
+              final Widget? shellNav =
```

Comment:
why does it grab the first child's navigator? shouldn't this shell route always created a navigator?

Reply:
This is changed now, but it was due to an additional level of nesting (additional ShellRoute) below.